### PR TITLE
Implement more of the API

### PR DESCRIPTION
**Pull request by rkuris**
Working up the stack to fill in some of the API methods and remove some dead commented-out code.

Fixed the last remaining benchmark, which required the testing grpc services, now working again.

Out of the box, there is less cache so it's slower than the old implementation. With a larger cache (20k nodes), it outperforms the old code (6s for 1M rows vs 84s).

Also fixed propose so it doesn't require a mutable DB.

### CODE REVIEW

**Review by richardpringle — APPROVED** [review]
I left two suggested changes because the compiler should automatically implement these traits, but when I went to double-check to see if things compile, I noticed quite a few issues.

Fixing these issues is not trivial. I spent about 20 minutes going down the rabbit hole, but every time I fixed all the compiler errors, new ones popped up. 

One example of something that I think is dangerous (in the code maintainability sense, not the runtime sense) is specialization through references. 

If you have a trait implementation as follows:
```rust
impl<T> Trait for TypeWrapper<T> where T: Trait {
    fn trait_fn(&self) -> ReturnType { /* ... */ }
}
```
You shouldn't also have a specific implementation 
```rust
impl Trait for &TypeWrapper<SomeSpecifcType> {
    fn trait_fn(&self) -> ReturnType { /* ... */ }
}
```
Using double-indirection to achieve specialization seems like it could make the code really difficult to modify. It's also unclear whether or not this was done purposefully.

HOWEVER, I know that none of this is a priority right now, just something to keep in the back of your mind and fix at some point in the future.

File: grpc-testtool/benches/insert.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ const TESTDIR: &str = "/tmp/benchdb";
 /// The port to use for testing
 const TESTPORT: u16 = 5000;
 /// The URI to connect to; this better match the TESTPORT
-const TESTURI: &str = "http://localhost:5000";
+const TESTURI: &str = "http://127.0.0.1:5000";
```
[the remark this answers is not in the document]
Comment (rkuris) [reply grpc-testtool/benches/insert.rs:23]: That works only for a listen address, not something to connect to.

https://www.techtarget.com/searchnetworking/answer/What-is-the-IP-address-0000-used-for#:~:text=On%20servers.-,IP%20address%200.0.,On%20PCs%20and%20client%20devices. has more information

File: storage/src/nodestore.rs [hunk 2]
```diff
@@ -587,6 +587,32 @@ pub trait HashedNodeReader: TrieReader {
...
+impl<T> TrieReader for Arc<T> where T: TrieReader {}
+
+impl<T> NodeReader for Arc<T>
+where
+    T: NodeReader,
+{
+    fn read_node(&self, addr: LinearAddress) -> Result<Arc<Node>, Error> {
+        T::read_node(self, addr)
+    }
+}
```
[the remark this answers is not in the document]
Comment (rkuris) [reply storage/src/nodestore.rs:605]: When I remove this, it won't compile, unfortunately.

[the remark this answers is not in the document]
Comment (rkuris) [reply storage/src/nodestore.rs]: This was resolved in another PR.

**Pull request merged**